components/accelerator/nvidia/ecc/component.go

```diff
@@ -3,156 +3,183 @@ package ecc
-		ms = append(ms, components.Metric{
-			Metric: m,
-			ExtraInfo: map[string]string{
-				"gpu_id": m.MetricSecondaryName,
-			},
-		})
-	}
-	for _, m := range aggTotalUncorrecteds {
-		ms = append(ms, components.Metric{
-			Metric: m,
-			ExtraInfo: map[string]string{
-				"gpu_id": m.MetricSecondaryName,
-			},
-		})
-	}
-	for _, m := range volTotalCorrecteds {
-		ms = append(ms, components.Metric{
-			Metric: m,
-			ExtraInfo: map[string]string{
-				"gpu_id": m.MetricSecondaryName,
-			},
-		})
+	return nil
+}
+
+// CheckOnce checks the current pods
+// run this periodically
+func (c *component) CheckOnce() {
+	log.Logger.Infow("checking ecc")
+	d := Data{
+		ts: time.Now().UTC(),
 	}
-	for _, m := range volTotalUncorrecteds {
-		ms = append(ms, components.Metric{
-			Metric: m,
-			ExtraInfo: map[string]string{
-				"gpu_id": m.MetricSecondaryName,
-			},
-		})
+	defer func() {
+		c.lastMu.Lock()
+		c.lastData = &d
+		c.lastMu.Unlock()
+	}()
+
+	devs := c.getDevicesFunc()
+	for uuid, dev := range devs {
+		eccMode, err := c.getECCModeEnabledFunc(uuid, dev)
+		if err != nil {
+			log.Logger.Errorw("error getting ECC mode for device", "uuid", uuid, "error", err)
+			d.err = err
+			d.healthy = false
+			d.reason = fmt.Sprintf("error getting ECC mode for device %s", uuid)
+			return
+		}
+		d.ECCModes = append(d.ECCModes, eccMode)
+
+		eccErrors, err := c.getECCErrorsFunc(uuid, dev, eccMode.EnabledCurrent)
+		if err != nil {
+			log.Logger.Errorw("error getting ECC errors for device", "uuid", uuid, "error", err)
+			d.err = err
+			d.healthy = false
+			d.reason = fmt.Sprintf("error getting ECC errors for device %s", uuid)
+			return
+		}
+
+		metricAggregateTotalCorrected.With(prometheus.Labels{pkgmetrics.MetricLabelKey: uuid}).Set(float64(eccErrors.Aggregate.Total.Corrected))
+		metricAggregateTotalUncorrected.With(prometheus.Labels{pkgmetrics.MetricLabelKey: uuid}).Set(float64(eccErrors.Aggregate.Total.Uncorrected))
+		metricVolatileTotalCorrected.With(prometheus.Labels{pkgmetrics.MetricLabelKey: uuid}).Set(float64(eccErrors.Volatile.Total.Corrected))
+		metricVolatileTotalUncorrected.With(prometheus.Labels{pkgmetrics.MetricLabelKey: uuid}).Set(float64(eccErrors.Volatile.Total.Uncorrected))
+
+		d.ECCErrors = append(d.ECCErrors, eccErrors)
 	}
 
-	return ms, nil
+	d.healthy = true
+	d.reason = fmt.Sprintf("checked %d devices for ECC errors", len(devs))
 }
 
-func (c *component) Close() error {
-	log.Logger.Debugw("closing component")
+type Data struct {
+	ECCModes  []nvidianvml.ECCMode   `json:"ecc_modes,omitempty"`
+	ECCErrors []nvidianvml.ECCErrors `json:"ecc_errors,omitempty"`
 
-	// safe to call stop multiple times
-	_ = c.poller.Stop(nvidia_ecc_id.Name)
+	// timestamp of the last check
+	ts time.Time `json:"-"`
+	// error from the last check
+	err error `json:"-"`
 
-	return nil
+	// tracks the healthy evaluation result of the last check
+	healthy bool `json:"-"`
+	// tracks the reason of the last check
+	reason string `json:"-"`
```

Comment:
Actually we don't need to set json tags for these fields since they are private, json package cannot access them. IDE like goland will report it as warning.